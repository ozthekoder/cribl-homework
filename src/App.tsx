import { useState, useEffect } from "react";
import { Log } from "./utils/types";
import { Column } from "./components/Column";
import { Row } from "./components/Row";
import { LogsTable } from "./components/LogsTable";
import { LOGS_URL } from "./utils/constants";
import { useLogs } from "./hooks/useLogs";
import LogsHistogram from "./components/LogsHistogram";

function App() {
  const { logs } = useLogs({ url: LOGS_URL });
  const histogramHeight = 200;
  const columns: (keyof Log)[] = ["Time", "Event"];

  // State to track available height
  const [tableHeight, setTableHeight] = useState(
    window.innerHeight - (histogramHeight + 32),
  );

  useEffect(() => {
    // Function to update height dynamically
    const handleResize = () => {
      setTableHeight(window.innerHeight - (histogramHeight + 32)); // Adjust table height dynamically
    };

    window.addEventListener("resize", handleResize); // Listen to window resize
    return () => window.removeEventListener("resize", handleResize); // Cleanup on unmount
  }, []);

  return (
    <Column>
      <Row height={`${histogramHeight}px`}>
        <LogsHistogram height={`${histogramHeight}px`} logs={logs} />
      </Row>
      <Row height={`calc(100% - ${histogramHeight}px)`}>
        <LogsTable
          rows={logs}
          columns={columns}
          height={tableHeight}
          rowHeight={32}
        />
      </Row>
    </Column>
  );
}

export default App;
