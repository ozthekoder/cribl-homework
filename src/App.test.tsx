/* eslint import/first: 0 */

import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";

jest.mock("./hooks/useLogs", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    logs: [
      {
        Time: "2024-03-01 12:00:00",
        Event: JSON.stringify({ message: "Log 0" }),
      },
      {
        Time: "2024-03-01 12:01:00",
        Event: JSON.stringify({ message: "Log 1" }),
      },
    ],
  })),
  useLogs: () => ({
    logs: [
      {
        Time: "2024-03-01 12:00:00",
        Event: JSON.stringify({ message: "Log 0" }),
      },
      {
        Time: "2024-03-01 12:01:00",
        Event: JSON.stringify({ message: "Log 1" }),
      },
    ],
  }),
}));

jest.mock("echarts-for-react", () => () => <div data-testid="mock-echarts" />);

import App from "./App";

describe("App Component", () => {
  test("should render without crashing", async () => {
    render(<App />);

    // Ensure the table headers are displayed
    expect(screen.getByText("Time")).toBeInTheDocument();
    expect(screen.getByText("Event")).toBeInTheDocument();

    // Ensure logs are rendered
    expect(screen.getByText("2024-03-01 12:00:00")).toBeInTheDocument();
    expect(screen.getByText("2024-03-01 12:01:00")).toBeInTheDocument();
  });

  test("should update table height on window resize", async () => {
    render(<App />);

    const table = screen.getByTestId("virtualized-container");

    window.innerHeight = 800;

    fireEvent(window, new Event("resize"));

    // Wait for re-render
    await screen.findByRole("table");

    // Ensure table height has been updated
    const computedStyle = window.getComputedStyle(table);
    expect(computedStyle.height).toBe("568px"); // 800 - (200 + 32 from useEffect calculation
  });

  test("should match snapshot", () => {
    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
