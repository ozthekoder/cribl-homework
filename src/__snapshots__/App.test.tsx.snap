// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match snapshot 1`] = `
<div
  class="css-m08lup"
>
  <div
    class="css-plwj0c"
    height="200px"
  >
    <div
      data-testid="mock-echarts"
    />
  </div>
  <div
    class="css-16g2hxh"
    height="calc(100% - 200px)"
  >
    <div
      class="css-10b489u"
      role="table"
    >
      <div
        class="css-1h2hj7p"
      >
        <div
          class="css-1vvyagg"
          width="200px"
        >
          Time
        </div>
        <div
          class="css-1e99wpz"
          width="calc(100% - 200px)"
        >
          Event
        </div>
      </div>
      <div
        class="css-10b489u"
        data-testid="virtualized-container"
        style="height: 568px;"
      >
        <div
          class="css-1eyo6l6"
        >
          <div
            class="css-9db80r"
            data-testid="virtualized-rows"
          >
            <div
              class="css-xspo3w"
              data-expanded="false"
              data-index="0"
              height="32"
              role="row"
            >
              <div
                class="css-ioidyq"
                width="200px"
              >
                <div
                  class="css-1li7dol"
                >
                  ❯
                </div>
                <div
                  class="css-vchpx1"
                >
                  2024-03-01 12:00:00
                </div>
              </div>
              <div
                class="css-16zk19m"
                width="calc(100% - 200px)"
              >
                <div
                  class="css-vchpx1"
                >
                  {"message":"Log 0"}
                </div>
              </div>
            </div>
            <div
              class="css-1hiimad"
              data-expanded="false"
              data-index="1"
              height="32"
              role="row"
            >
              <div
                class="css-ioidyq"
                width="200px"
              >
                <div
                  class="css-1li7dol"
                >
                  ❯
                </div>
                <div
                  class="css-vchpx1"
                >
                  2024-03-01 12:01:00
                </div>
              </div>
              <div
                class="css-16zk19m"
                width="calc(100% - 200px)"
              >
                <div
                  class="css-vchpx1"
                >
                  {"message":"Log 1"}
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
